configserver/src/main/java/com/yahoo/vespa/config/server/ConfigServerBootstrap.java

```diff
@@ -3,56 +3,70 @@
 import com.yahoo.log.LogLevel;
 import com.yahoo.vespa.config.server.rpc.RpcServer;
 import com.yahoo.vespa.config.server.version.VersionState;
 
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
 /**
  * Main component that bootstraps and starts config server threads.
  *
- * @author lulf
- * @since 5.1
+ * If config server has been upgraded to a new version since the last time it was running it will redeploy all
+ * applications. If that is done syúccessfully the RPC server will start and the health status code will change from
+ * 'initializing' to 'up' and the config server will be put into rotation (start serving status.html with 200 OK)
+ *
+ * @author Ulf Lilleengen
+ * @author hmusum
  */
 public class ConfigServerBootstrap extends AbstractComponent implements Runnable {
 
     private static final java.util.logging.Logger log = java.util.logging.Logger.getLogger(ConfigServerBootstrap.class.getName());
+    private static final ExecutorService rpcServerExecutor  = Executors.newSingleThreadExecutor(new DaemonThreadFactory("config server RPC server"));
+    private static final String vipStatusClusterIdentifier = "configserver";
 
     private final ApplicationRepository applicationRepository;
     private final RpcServer server;
     private final Thread serverThread;
     private final VersionState versionState;
     private final StateMonitor stateMonitor;
+    private final VipStatus vipStatus;
 
     // The tenants object is injected so that all initial requests handlers are
     // added to the rpc server before it starts answering rpc requests.
     @SuppressWarnings("WeakerAccess")
     @Inject
     public ConfigServerBootstrap(ApplicationRepository applicationRepository, RpcServer server,
-                                 VersionState versionState, StateMonitor stateMonitor) {
-        this(applicationRepository, server, versionState, stateMonitor, true);
+                                 VersionState versionState, StateMonitor stateMonitor, VipStatus vipStatus) {
+        this(applicationRepository, server, versionState, stateMonitor, vipStatus, true);
     }
 
     // For testing only
-    ConfigServerBootstrap(ApplicationRepository applicationRepository, RpcServer server,
-                          VersionState versionState, StateMonitor stateMonitor, boolean startMainThread) {
+    ConfigServerBootstrap(ApplicationRepository applicationRepository, RpcServer server, VersionState versionState,
+                          StateMonitor stateMonitor, VipStatus vipStatus, boolean startMainThread) {
         this.applicationRepository = applicationRepository;
         this.server = server;
         this.versionState = versionState;
         this.stateMonitor = stateMonitor;
         this.serverThread = new Thread(this, "configserver main");
+        this.vipStatus = vipStatus;
+        initializing(); // Initially take server out of rotation
```

Comment:
VipStatus.inRotation() is unfortunately true by default, so there will be a window where status.html says it's up and before we reach this point.

It would be nice to use a "DefaultOutOfRotationVipStatus" instead of VipStatus. Or add an initialInRotation (default true) to VipStatusConfig and have that injected to VipStatus? Third alternative is to use the file.